bioio/writers/ome_zarr_writer_2.py

```diff
@@ -0,0 +1,688 @@
+import logging
+from dataclasses import asdict, dataclass
+from math import prod
+from typing import Any, List, Tuple, Union
+
+import dask.array as da
+import numcodecs
+import numpy as np
+import skimage.transform
+import zarr
+from ngff_zarr.zarr_metadata import Axis, Dataset, Metadata, Scale, Translation
+from zarr.storage import DirectoryStore, FSStore, default_compressor
+
+from bioio import BioImage
+
+log = logging.getLogger(__name__)
+
+DimTuple = Tuple[int, int, int, int, int]
+
+
+def chunk_size_from_memory_target(
+    shape: DimTuple, dtype: str, memory_target: int
+) -> DimTuple:
+    """
+    Calculate chunk size from memory target in bytes.  The chunk size will be
+    determined by considering a single T and C, and subdividing the remaining
+    dims by 2 until the chunk fits within the size target.
+
+    Parameters
+    ----------
+    shape:
+        Shape of the array. Assumes a 5d TCZYX array.
+    dtype:
+        Data type of the array.
+    memory_target:
+        Memory target in bytes.
+
+    Returns
+    -------
+    Chunk size tuple.
+    """
+    if len(shape) != 5:
+        raise ValueError("shape must be a 5-tuple in TCZYX order")
+
+    itemsize = np.dtype(dtype).itemsize
+    # let's start by just mandating that chunks have to be no more than
+    # 1 T and 1 C
+    chunk_size = (1, 1, shape[2], shape[3], shape[4])
+    while prod(chunk_size) * itemsize > memory_target:
+        # chop every dim in half until they get down to 1
+        chunk_size = (
+            max(chunk_size[0] // 2, 1),
+            max(chunk_size[1] // 2, 1),
+            max(chunk_size[2] // 2, 1),
+            max(chunk_size[3] // 2, 1),
+            max(chunk_size[4] // 2, 1),
+        )
+    return chunk_size
+
+
+def dim_tuple_to_dict(
+    dims: Union[DimTuple, Tuple[float, float, float, float, float]]
+) -> dict:
+    if len(dims) != 5:
+        raise ValueError("dims must be a 5-tuple in TCZYX order")
+    return {"t": dims[0], "c": dims[1], "z": dims[2], "y": dims[3], "x": dims[4]}
+
+
+def resize(
+    image: da.Array, output_shape: Tuple[int, ...], *args: Any, **kwargs: Any
+) -> da.Array:
+    r"""
+    Wrapped copy of "skimage.transform.resize"
+    Resize image to match a certain size.
+
+    Parameters
+    ----------
+    image: :class:`dask.array`
+        The dask array to resize
+    output_shape: tuple
+        The shape of the resize array
+    \*args: list
+        Arguments of skimage.transform.resize
+    \*\*kwargs: dict
+        Keyword arguments of skimage.transform.resize
+
+    Returns
+    -------
+    Resized image.
+    """
+    factors = np.array(output_shape) / np.array(image.shape).astype(float)
+    # Rechunk the input blocks so that the factors achieve an output
+    # blocks size of full numbers.
+    better_chunksize = tuple(
+        np.maximum(1, np.round(np.array(image.chunksize) * factors) / factors).astype(
```

Comment:
nit. Different methods of finding maximum throughout file

Reply:
I'm not going to modify this one as it is very specific for the downsampling code and is borrowed from a trusted implementation